File: src/doppler/examples/async_dsss_pool_demo.py
"""async_dsss_pool_demo.py -- the multi-emitter pool as ONE object: two
emitters on one code arrive, one leaves and is released, it returns and
is a new detection -- the lifecycle :class:`~doppler.dsss.AsyncDsssPool`
holds (``docs/design/async-dsss-receiver.md`` section 8.2), with the
event log attached and a timeline of who held which slot when.

The C twin (``native/examples/async_dsss_pool_demo.c``) manages the
lifecycle the binding hides -- the symbol buffer, the borrowed log, the
slot keyed by both coordinates. This page shows the *result*: the slots'
occupancy against the truth, each tracked receiver's live Doppler, and
the log's transitions in order.

**The waveform is not built here.** Each emitter is the shipped
continuous-DSSS :class:`~doppler.wfm.Synth` (the code from
:class:`~doppler.wfm.Gold`, asynchronous BPSK from its own seeded PN) at
its own carrier offset, and the noise is a ``type="noise"`` Synth scaled
by :func:`~doppler.wfm.wfm_awgn_amplitude` from the C/N0 -- nothing here
spreads a chip, draws a bit or a sigma.

**A slot is the emitter's by both coordinates, never by a count.** The
searcher's false alarms are part of the lifecycle: at pfa 1e-3 a noise
peak seeds a free slot every few hundred milliseconds, refines to
nothing and is released one interval later. So an emitter's slot is the
one whose SEED is within the searcher's row of its Doppler AND within a
chip of its code phase at the seed's sample -- the synth's own clock --
and the figure colours every other assignment as what it is.
"""

from __future__ import annotations

import json
import os
import sys
import tempfile

# --8<-- [start:geometry]
import numpy as np

from doppler.dsss import AsyncDsssPool
from doppler.telemetry import EventLog
from doppler.wfm import Gold, Synth, wfm_awgn_amplitude

SF = 1023
CHIP_RATE = 5.0e6
SPC = 2
FS = CHIP_RATE * SPC
SYM_RATE = 2700.0
TE = SF * SPC  # one code epoch, the push block
CN0_DBHZ = 47.0
DU = 6000.0  # the searcher's span: +-6 kHz, D = 1 rows of 4.9 kHz
LOST_S = 0.3  # the release interval, short enough to see
N_SLOTS = 4
EMITTERS = {  # name: (Doppler Hz, code phase at sample 0 in chips, seed)
    "A": (1500.0, 0, 1),
    "B": (-3500.0, 900, 2),
}
ON_S, OFF_S = 1.0, 3.0 * LOST_S  # A always on; B on, off, on
CODE = np.asarray(Gold().generate(SF)).astype(np.uint8)
# --8<-- [end:geometry]


# --8<-- [start:stimulus]
def emitter(doppler_hz: float, chip0: int, seed: int) -> Synth:
    """One emitter: the shipped continuous DSSS at a carrier offset, clean,
    its code `chip0` chips in at the stream's first sample (a burn-in the
    caller discards) -- two emitters must differ in code phase as well as
    Doppler, since one phase is one peak to the searcher."""
    syn = Synth(
        type="dsss",
        data_code=bytes(CODE.tolist()),
        symbol_rate=SYM_RATE,
        sps=SPC,
        snr=100.0,  # clean: the noise is added once, at the sum
        fs=FS,
        freq=doppler_hz,
        seed=seed,
    )
    if chip0:
        syn.steps(chip0 * SPC)
    return syn


def stimulus(seed: int = 7):
    """Both emitters on for ON_S, B off for OFF_S (its synth keeps running:
    coming back is not restarting), both on again; noise at the sum.
    Returns the capture and B's on-air mask per epoch."""
    a = emitter(*EMITTERS["A"])
    b = emitter(*EMITTERS["B"])
    n_on, n_off = int(ON_S * FS / TE), int(OFF_S * FS / TE)
    on_b = np.array([1] * n_on + [0] * n_off + [1] * n_on, dtype=bool)
    blocks = []
    for on in on_b:
        sa = a.steps(TE)
        sb = b.steps(TE)
        blocks.append(sa + sb if on else sa)
    x = np.concatenate(blocks)
    # C/N0 referred to fs; the Synth emits unit total power, the amplitude
    # is per component, and the sqrt(2) bridges the two conventions.
    amp = float(wfm_awgn_amplitude(CN0_DBHZ - 10.0 * np.log10(FS), 1.0))
    noise = Synth(type="noise", fs=FS, seed=seed).steps(len(x))
    return (x + amp * np.sqrt(2.0) * noise).astype(np.complex64), on_b


# --8<-- [end:stimulus]


# --8<-- [start:pool]
def owner(pool: AsyncDsssPool, slot: int) -> str | None:
    """Which emitter a slot's SEED is -- row and chip phase -- or None
    for a false alarm (a noise seed at another phase)."""
    r = pool.status(slot)
    if not r.assigned:
        return None
    for name, (f, chip0, _) in EMITTERS.items():
        truth = (r.seed_sample / SPC + chip0) % SF
        dc = abs(r.seed_chip_phase - truth)
        dc = min(dc, SF - dc)
        if abs(r.seed_doppler_hz - f) <= pool.doppler_res_hz and dc <= 1.0:
            return name
    return "false alarm"


def run(x: np.ndarray, events_path: str):
    pool = AsyncDsssPool(
        CODE,
        chip_rate=CHIP_RATE,
        symbol_rate=SYM_RATE,
        spc=SPC,
        cn0_dbhz=CN0_DBHZ,
        pfa=1e-3,
        doppler_uncertainty=DU,
        n_slots=N_SLOTS,
        lost_confirm_s=LOST_S,
        threads=1,
    )
    log = EventLog(events_path)
    pool.set_event_log(log)
    n_ep = len(x) // TE
    who = np.full((N_SLOTS, n_ep), "", dtype=object)  # owner per epoch
    dopp = np.full((N_SLOTS, n_ep), np.nan)  # live Doppler while tracking
    n_syms = dict.fromkeys(EMITTERS, 0)
    for k in range(n_ep):
        pool.push(x[k * TE : (k + 1) * TE])
        for i in range(N_SLOTS):
            o = owner(pool, i)
            r = pool.status(i)
            who[i, k] = o or ""
            if o in n_syms:
                n_syms[o] += len(pool.symbols(i))
            if r.assigned and r.state == 2:  # tracking
                dopp[i, k] = r.doppler_hz
    pool.set_event_log(None)
    log.close()
    with open(events_path) as fh:
        events = [json.loads(line) for line in fh]
    assert len(events) == pool.events, "every transition reached the log"
    return who, dopp, n_syms, events


# --8<-- [end:pool]


def check(who, on_b, n_syms, events) -> None:
    """The lifecycle, asserted against the stimulus's truth."""
    t = np.arange(who.shape[1]) * TE / FS
    held = {n: (who == n).sum(axis=0) for n in EMITTERS}  # slots per epoch
    # Assigned once: never two slots on one emitter.
    for n in EMITTERS:
        assert held[n].max() == 1, f"{n} was held by two slots at once"
    # A: tracked for the whole run once found; B: found in each stint.
    first_a = np.argmax(held["A"] > 0)
    assert held["A"][first_a:].mean() > 0.99, "A was not held throughout"
    off = np.flatnonzero(~on_b)
    t_off, t_back = t[off[0]], t[off[-1] + 1]
    b_before = held["B"][: off[0]]
    assert b_before.max() == 1, "B was never assigned in its first stint"
    # Released by the rule after the departure -- inside the interval plus
    # half a second, never before it -- and a new slot after the return.
    rel = np.flatnonzero(held["B"][off[0] :] == 0)
    assert len(rel), "B's slot was never released"
    t_rel = t[off[0] + rel[0]] - t_off
    assert LOST_S <= t_rel <= LOST_S + 0.5, f"released at {t_rel:.2f} s"
    slot_before = int(np.flatnonzero(who[:, off[0] - 1] == "B")[0])
    back = np.flatnonzero(held["B"][off[-1] + 1 :] > 0)
    assert len(back), "B was not re-acquired on return"
    slot_after = int(np.flatnonzero(who[:, off[-1] + 1 + back[0]] == "B")[0])
    assert n_syms["A"] > 1000 and n_syms["B"] > 1000, "symbols by slot"
    labels = [e["core:label"] for e in events]
    want = ["seeded", "tracking", "lost", "released", "seeded"]
    i = 0
    for lab in labels:
        if i < len(want) and lab == want[i]:
            i += 1
    assert i == len(want), f"the log's order: {labels}"
    print(
        f"A: one slot from {t[first_a]:.3f} s to the end; "
        f"B: slot {slot_before} until it left at {t_off:.2f} s, released "
        f"{t_rel * 1e3:.0f} ms later (the interval is {LOST_S * 1e3:.0f} ms), "
        f"slot {slot_after} after it returned at {t_back:.2f} s; "
        f"{n_syms['A']} + {n_syms['B']} symbols; {len(events)} events"
    )


def figure(who, dopp, on_b, out_path: str) -> None:
    import matplotlib

    matplotlib.use("Agg")
    import matplotlib.pyplot as plt

    t = np.arange(who.shape[1]) * TE / FS
    colours = {"A": "tab:blue", "B": "tab:orange", "false alarm": "tab:gray"}
    fig, (ax0, ax1) = plt.subplots(
        2, 1, figsize=(9, 5.5), sharex=True, constrained_layout=True
    )
    seen: set[str] = set()
    for i in range(N_SLOTS):
        for name, c in colours.items():
            m = who[i] == name
            if not m.any():
                continue
            ax0.fill_between(
                t,
                i - 0.4,
                i + 0.4,
                where=m,
                color=c,
                step="mid",
                label=None if name in seen else name,
            )
            seen.add(name)
        ax1.plot(t, dopp[i], lw=1, label=f"slot {i}")
    off = np.flatnonzero(~on_b)
    for ax in (ax0, ax1):
        ax.axvline(t[off[0]], color="k", ls="--", lw=0.8)
        ax.axvline(t[off[-1] + 1], color="k", ls="--", lw=0.8)
    ax0.set_yticks(range(N_SLOTS))
    ax0.set_ylabel("slot")
    ax0.set_title(
        "AsyncDsssPool: who holds which slot -- B leaves at the first "
        "dashed line, returns at the second"
    )
    ax0.legend(loc="upper right", fontsize=8)
    for f, _, _ in EMITTERS.values():
        ax1.axhline(f, color="k", lw=0.5, alpha=0.4)
    ax1.set_ylabel("live Doppler (Hz)")
    ax1.set_xlabel("time (s)")
    ax1.legend(loc="upper right", fontsize=8)
    fig.savefig(out_path, dpi=120)
    plt.close(fig)


def main(out_path: str = "async_dsss_pool_demo.png") -> None:
    x, on_b = stimulus()
    with tempfile.TemporaryDirectory() as d:
        who, dopp, n_syms, events = run(x, os.path.join(d, "run.events"))
    check(who, on_b, n_syms, events)
    figure(who, dopp, on_b, out_path)
    print(f"wrote {out_path}")


if __name__ == "__main__":
    main(*sys.argv[1:2])

File: native/examples/async_dsss_pool_demo.c
/**
 * async_dsss_pool_demo.c — the multi-emitter pool as ONE object, in C.
 *
 * `AsyncDsssPool` holds the population of the continuous async-DSSS use
 * case (docs/design/async-dsss-receiver.md §8.2): one searcher, `n_slots`
 * hand-off receivers created idle, the assigned table and the event log,
 * behind a single `push()`. The Python twin
 * (`src/doppler/examples/async_dsss_pool_demo.py`) shows the same
 * lifecycle with a figure; this one shows what the binding does FOR you
 * and therefore hides: the lifecycle you manage yourself, who owns the
 * symbol buffer and how big it has to be, and the log you attach and
 * close.
 *
 * **The waveform is not built here.** Each emitter is the shipped
 * `wfm_synth` continuous DSSS -- the code from wfm's own Gold generator,
 * asynchronous BPSK from its own PRBS -- at its own carrier offset, and
 * the noise is the shipped awgn sized from the C/N0. Nothing here spreads
 * a chip, draws a bit or a sigma: a second implementation of any of those
 * would drift from the one that ships.
 *
 * Four sections, each printing a number a reader can check:
 *
 *   §1  Two emitters on one code, two rows apart: each takes exactly one
 *       slot, its seed within the searcher's row of its own Doppler; both
 *       track with code lock; their symbols come out by slot.
 *   §2  One leaves the air. Its receiver reports lost once both lock
 *       flags have been down for the release interval, and the pool
 *       releases the slot -- the other emitter is untouched.
 *   §3  It returns: a new detection into whichever slot is free -- the
 *       one re-assignment the lifecycle permits.
 *   §4  The event log, attached by borrowing: every transition the pool
 *       counted is one appended line, in order.
 *
 * Build:
 *   cmake --build build
 *   ./build/native/examples/async_dsss_pool_demo
 */
#include <async_dsss_pool/async_dsss_pool_core.h>
#include <awgn/awgn_core.h>
#include <complex.h>
#include <dp_event_log/dp_event_log_core.h>
#include <gold/gold_core.h>
#include <math.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>
#include <wfm_synth/wfm_synth_core.h>

/* ── geometry: the operating point of design §6.1, at D = 1 ─────────────
 * 1023 chips at 5 Mcps, two samples per chip, 2700 sym/s asynchronous BPSK.
 * The searcher runs epoch by epoch (no code-only window in this stream, so
 * D = 1) over ±6 kHz: a Doppler row is one epoch rate, 4.89 kHz, and two
 * emitters two rows apart are two peaks. The release interval is short
 * so a departure is released inside a second of stream. */
#define SF 1023u
#define SPC 2u
#define TE (SF * SPC)
#define CHIP_RATE 5.0e6
#define FS (CHIP_RATE * (double)SPC)
#define SYM_RATE 2700.0
#define CN0_DBHZ 47.0
#define DU 6000.0
#define LOST_S 0.3
#define N_SLOTS 4u
#define DOPPLER_A 1500.0
#define DOPPLER_B (-3500.0)
#define CHIP0_B 900.0 /* B's code phase at sample 0: its burn-in, chips */

#define CHECK(cond, ...)                                                      \
  do                                                                          \
    {                                                                         \
      if (!(cond))                                                            \
        {                                                                     \
          fprintf (stderr, "FAIL: " __VA_ARGS__);                             \
          fputc ('\n', stderr);                                               \
          return 1;                                                           \
        }                                                                     \
    }                                                                         \
  while (0)

/** @brief One emitter: the shipped continuous-DSSS synth at a carrier
 *  offset, its own PRBS data, clean -- the noise is added once at the sum
 *  so two emitters share one channel's noise. */
static wfm_synth_state_t *
emitter (const uint8_t *code, double doppler_hz, uint32_t seed)
{
  wfm_synth_state_t *syn
      = wfm_synth_create (WFM_SYNTH_DSSS, FS, doppler_hz, WFM_SYNTH_SNR_CLEAN,
                          1, seed, (int)SPC, 15, 0, 0, 0.0);
  if (syn
      && wfm_synth_set_dsss_cont (syn, code, SF, CHIP_RATE / SYM_RATE,
                                  WFM_DSSS_DATA_PRBS, NULL, 0)
             != 0)
    {
      wfm_synth_destroy (syn);
      syn = NULL;
    }
  return syn;
}

/** @brief The slot whose SEED is the emitter at @p doppler_hz -- within
 *  the searcher's row AND within a chip of the emitter's own code phase at
 *  the seed's sample (the synth's chip clock, `chip0` chips in at sample
 *  0) -- or N_SLOTS; and how many slots hold it. The searcher's false
 *  alarms are part of the lifecycle: a noise peak in the same row seeds a
 *  free slot for one release interval, at another phase, so a slot is the
 *  emitter's by both coordinates, never by a count. */
static size_t
slot_of (async_dsss_pool_state_t *p, double doppler_hz, double chip0,
         size_t *count)
{
  size_t found = N_SLOTS, n = 0;
  for (size_t i = 0; i < N_SLOTS; i++)
    {
      async_dsss_pool_slot_t r = async_dsss_pool_status (p, i);
      if (!r.assigned
          || fabs (r.seed_doppler_hz - doppler_hz) > p->acq->doppler_res_hz)
        continue;
      double truth = dp_fmod_pos ((double)r.seed_sample / SPC + chip0, SF);
      double dc    = fabs (r.seed_chip_phase - truth);
      if (dc > SF / 2.0)
        dc = SF - dc;
      if (dc <= 1.0)
        {
          found = i;
          n++;
        }
    }
  *count = n;
  return found;
}

int
main (void)
{
  uint8_t       code[SF];
  gold_state_t *gd = gold_create (934, 350, 567, 73, 10);
  gold_generate (gd, SF, code, SF);
  gold_destroy (gd);

  wfm_synth_state_t *a = emitter (code, DOPPLER_A, 1u);
  wfm_synth_state_t *b = emitter (code, DOPPLER_B, 2u);
  awgn_state_t      *g = awgn_create (
      7u,
      awgn_amplitude_for_snr ((float)(CN0_DBHZ - 10.0 * log10 (FS)), 1.0f));
  CHECK (a && b && g, "the emitters and the noise open");
  /* B starts 900 chips into its code: two emitters at one code phase are
     one peak to the searcher and one row to the pool's zone (§7.1 -- a
     pair the surface cannot tell apart), so they must differ in phase as
     well as in Doppler, as real emitters do. The burn-in is discarded. */
  {
    float complex *burn = malloc ((size_t)CHIP0_B * SPC * sizeof *burn);
    CHECK (burn != NULL, "the burn-in allocates");
    wfm_synth_steps (b, burn, (size_t)CHIP0_B * SPC);
    free (burn);
  }

  /* The pool: the searcher's and the receivers' parameters pass through
     create() untouched; everything is sized here, once. */
  async_dsss_pool_state_t *pool = async_dsss_pool_create (
      code, SF, CHIP_RATE, SYM_RATE, SPC, 2, CN0_DBHZ, 1e-3, 0.9, DU, 1, 0.0,
      4, N_SLOTS, 1, 0.0, LOST_S, 0.0, 4, 8, 0, 0.5, 4, 14.0, 64, 8, false,
      100000);
  CHECK (pool != NULL, "the pool opens");

  /* §4's log, attached before the first push so nothing is missed. It is
     borrowed: the pool never closes it. */
  char path[64];
  (void)snprintf (path, sizeof path, "/tmp/async_dsss_pool_demo_%d.events",
                  (int)getpid ());
  dp_event_log_t *log = dp_event_log_open (path, 0.0);
  CHECK (log != NULL && async_dsss_pool_set_event_log (pool, log) == DP_OK,
         "the event log opens and attaches");

  float complex *sa  = malloc ((size_t)TE * sizeof *sa);
  float complex *sb  = malloc ((size_t)TE * sizeof *sb);
  float complex *x   = malloc ((size_t)TE * sizeof *x);
  size_t         cap = 0; /* the symbol buffer's capacity: known only after
                             the first push, since it follows the block */
  float complex *syms            = NULL;
  size_t         n_syms[N_SLOTS] = { 0 };
  CHECK (sa && sb && x, "the blocks allocate");

  /* One block: both emitters (b's gain is its visibility), plus noise. */
#define BLOCK(gain_b)                                                         \
  do                                                                          \
    {                                                                         \
      wfm_synth_steps (a, sa, TE);                                            \
      wfm_synth_steps (b, sb, TE);                                            \
      awgn_generate (g, TE, x, TE);                                           \
      for (size_t i = 0; i < TE; i++)                                         \
        x[i] += sa[i] + (float)(gain_b) * sb[i];                              \
    }                                                                         \
  while (0)

  /* ── §1 ───────────────────────────────────────────────────────────── */
  printf ("§1  two emitters, one code, %.0f and %.0f Hz\n", DOPPLER_A,
          DOPPLER_B);
  const size_t on_blocks = (size_t)(1.0 * FS / TE); /* one second */
  for (size_t k = 0; k < on_blocks; k++)
    {
      BLOCK (1.0);
      (void)async_dsss_pool_push (pool, x, TE);
      if (!syms)
        {
          /* Sized by the pool from the largest block it has seen; a
             caller reads it after the first push and keeps it. */
          cap  = async_dsss_pool_symbols_max_out (pool);
          syms = malloc (cap * sizeof *syms);
          CHECK (syms != NULL, "the symbol buffer allocates");
        }
      for (size_t i = 0; i < N_SLOTS; i++)
        n_syms[i] += async_dsss_pool_symbols (pool, i, syms, cap);
    }
  size_t na, nb;
  size_t slot_a = slot_of (pool, DOPPLER_A, 0.0, &na);
  size_t slot_b = slot_of (pool, DOPPLER_B, CHIP0_B, &nb);
  CHECK (na == 1 && nb == 1 && slot_a != slot_b,
         "each emitter holds exactly one slot (A %zu, B %zu)", na, nb);
  async_dsss_pool_slot_t ra = async_dsss_pool_status (pool, slot_a);
  async_dsss_pool_slot_t rb = async_dsss_pool_status (pool, slot_b);
  printf ("    A: slot %zu, seeded at %.3f s at %+.0f Hz, now %+.1f Hz, "
          "state %d, code lock %d, symbol lock %d, %zu symbols so far\n",
          slot_a, (double)ra.seed_sample / FS, ra.seed_doppler_hz,
          ra.doppler_hz, ra.state, ra.code_locked, ra.locked, n_syms[slot_a]);
  printf ("    B: slot %zu, seeded at %.3f s at %+.0f Hz, now %+.1f Hz, "
          "state %d, code lock %d, symbol lock %d, %zu symbols so far\n",
          slot_b, (double)rb.seed_sample / FS, rb.seed_doppler_hz,
          rb.doppler_hz, rb.state, rb.code_locked, rb.locked, n_syms[slot_b]);
  CHECK (ra.state == ASYNC_DSSS_RX_TRACKING && ra.code_locked
             && rb.state == ASYNC_DSSS_RX_TRACKING && rb.code_locked,
         "both track their emitter with code lock");
  CHECK (fabs (ra.doppler_hz - DOPPLER_A) < 100.0
             && fabs (rb.doppler_hz - DOPPLER_B) < 100.0,
         "the live Doppler is each emitter's own");
  CHECK (n_syms[slot_a] > 1000 && n_syms[slot_b] > 1000,
         "symbols come out by slot");
  CHECK (pool->dropped == 0, "nothing dropped with slots to spare");

  /* ── §2 ───────────────────────────────────────────────────────────── */
  printf ("§2  B leaves the air\n");
  const size_t off_blocks  = (size_t)(3.0 * LOST_S * FS / TE);
  size_t       released_at = 0;
  for (size_t k = 0; k < off_blocks; k++)
    {
      BLOCK (0.0);
      (void)async_dsss_pool_push (pool, x, TE);
      if (!released_at && !async_dsss_pool_status (pool, slot_b).assigned)
        released_at = k + 1;
    }
  CHECK (released_at > 0, "B's slot is released once it is gone");
  printf ("    released %.0f ms after the departure (the interval is %.0f "
          "ms); A still tracking: %d\n",
          (double)released_at * TE / FS * 1e3, LOST_S * 1e3,
          async_dsss_pool_status (pool, slot_a).state
              == ASYNC_DSSS_RX_TRACKING);
  CHECK ((double)released_at * TE / FS >= LOST_S,
         "not before the release interval has run");
  CHECK (async_dsss_pool_status (pool, slot_a).state == ASYNC_DSSS_RX_TRACKING,
         "A is untouched by B's release");

  /* ── §3 ───────────────────────────────────────────────────────────── */
  printf ("§3  B returns\n");
  for (size_t k = 0; k < on_blocks; k++)
    {
      BLOCK (1.0);
      (void)async_dsss_pool_push (pool, x, TE);
    }
  size_t slot_b2 = slot_of (pool, DOPPLER_B, CHIP0_B, &nb);
  CHECK (nb == 1 && slot_b2 < N_SLOTS,
         "B is a new detection into a free slot");
  printf ("    B: slot %zu (was %zu), state %d; A: still slot %zu, %d\n",
          slot_b2, slot_b, async_dsss_pool_status (pool, slot_b2).state,
          slot_a, async_dsss_pool_status (pool, slot_a).state);
  CHECK (async_dsss_pool_status (pool, slot_b2).state
             == ASYNC_DSSS_RX_TRACKING,
         "and tracks again");

  /* ── §4 ───────────────────────────────────────────────────────────── */
  printf ("§4  the event log\n");
  CHECK (dp_event_log_close (log) == DP_OK, "the log closes");
  CHECK (dp_event_log_count (log) == pool->events,
         "every transition the pool counted reached the log (%zu of %llu)",
         dp_event_log_count (log), (unsigned long long)pool->events);
  static const char *const want[] = { "\"seeded\"", "\"tracking\"", "\"lost\"",
                                      "\"released\"", "\"seeded\"" };
  FILE                    *f      = fopen (path, "r");
  size_t                   found  = 0;
  char                     line[DP_EVENT_LOG_LINE_MAX];
  CHECK (f != NULL, "the flat file reads back");
  while (found < 5 && fgets (line, sizeof line, f))
    if (strstr (line, want[found]))
      found++;
  fclose (f);
  printf ("    %zu lines; seeded -> tracking -> lost -> released -> seeded, "
          "in order: %s\n",
          dp_event_log_count (log), found == 5 ? "yes" : "NO");
  CHECK (found == 5, "the labels are the design's, in order");
  (void)async_dsss_pool_set_event_log (pool, NULL);
  dp_event_log_destroy (log);
  remove (path);

  free (syms);
  free (x);
  free (sb);
  free (sa);
  async_dsss_pool_destroy (pool);
  awgn_destroy (g);
  wfm_synth_destroy (b);
  wfm_synth_destroy (a);
  printf ("OK\n");
  return 0;
}
